### ISSUE TITLE
Group 07 - chemboxR

### ISSUE THREAD

**Issue by Suraporn** [UBC-MDS/software-review-2023#8]
---
name: chemboxR
about: Molecular information calculator based on empirical formulas of chemicals in raw text.

---

Submitting Author Name: Wilfred Hass, Vikram Grewal, Luke Yang, NatePuangpanbut.
Submitting Author Github Handle: @WilfHass, @xFiveRivers, @lukeyf, @Suraporn 

Repository: https://github.com/UBC-MDS/chemboxR
Version submitted: 0.2.0
Submission type: Standard
Editor:  Wilfred Hass (@WilfHass), Vikram Grewal (@xFiveRivers), Luke Yang (@lukeyf), NatePuangpanbut (@Suraporn). 
Reviewers:  Mehwish, Robin Dhillon, Flora Ouedraogo, Yukun Zhang 

Archive: TBD
Version accepted: TBD
Language: en
---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: chemboxR
Title: ChemboxR is an R copy of Chembox, originally written in python
Version: 0.0.0.9000
Authors@R: 
    c(person("Wilfred", "Hass", , "wthass@student.ubc.ca", role = c("aut", "cre")),
    person("Luke", "Yang", , "lukey.ubc@qq.com", role = c("aut")),
    person("Nate", "Suraporn", , "", role = c("aut")),
    person("Vikram", "Grewal", , "", role = c("aut")))
Description: `chemboxR` is a package for molecular information calculator based 
    on empirical formulas of chemicals in raw text. It is designed to 
    intelligently process text input containing the chemical formula and provide 
    associated information on the inquired molecule. It is able to provide the 
    molar mass, check a formula's validity, and provide a balanced combustion 
    equation if the input is combustible. This tool can be utilized for various 
    educational and research purposes for simple and fast information retrieval.
License: MIT + file LICENSE
Encoding: UTF-8
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.2.3
Imports:
    tidyverse
Suggests: 
    testthat (>= 3.0.0)
Config/testthat/edition: 3
```

## Scope

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [ ] data retrieval
	- [ ] data extraction
	- [ ] data munging
	- [ ] data deposition
    - [ ] data validation and testing
	- [ ] workflow automation
	- [ ] version control
	- [ ] citation management and bibliometrics
	- [x] scientific software wrappers
	- [ ] field and lab reproducibility tools
	- [ ] database software bindings
	- [ ] geospatial data
	- [ ] text analysis

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):
`chemboxR` provides useful information extracted from a text chemical formula. Therefore, it is a wrapper functionfor Chemistry application. 

-   Who is the target audience and what are scientific applications of this package?
The target audience is anyone who is interested in extracting chemical properties from a chemical formula. It can high school students, undergraduate students, or any chemistry enthusiastic people who want to learn about chemical formula properties.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?
A similar package [chemsolve](https://github.com/amogh7joshi/chemsolve) is available online that employs a similar string-parsing design and molar mass calculation. Yet, we include a dataframe of the molecule properties to give the user enhanced functionality and insights. In addition, we include methods for automated combustion equation generation and balancing.

-   (If applicable) Does your package comply with our [guidance around _Ethics, Data Privacy and Human Subjects Research_](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?

-   If you made a pre-submission inquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

-   Explain reasons for any [`pkgcheck` items](https://docs.ropensci.org/pkgcheck/) which your package is unable to pass.

## Technical checks

Confirm each of the following by checking the box.

- [x] I have read the [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).
- [x] I have read the [author guide](https://devdevguide.netlify.app/authors-guide.html) and I expect to maintain this package for at least 2 years or to find a replacement.

This package:

- [x] does not violate the Terms of Service of any service it interacts with.
- [ ] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme).
- [x] includes [documentation with examples for all functions, created with roxygen2](https://ropensci.github.io/dev_guide/building.html#documentation).
- [ ] contains a vignette with examples of its essential functions and uses.
- [x] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [ ] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage.

## Publication options

- [ ] Do you intend for this package to go on CRAN?
- [ ] Do you intend for this package to go on Bioconductor?

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:

<details>
<summary>MEE Options</summary>

- [ ] The package is novel and will be of interest to the broad readership of the journal.
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.org/code-of-conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by florawendy19** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples:** (that run successfully locally) for all exported functions
- [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing:

1 hour

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

 I was able to install the package without any difficulty and the different functions work perfectly and the outputs are the expected ones.

In the R package, the different functions are in different R files unlike in Python and it makes it easy to review the code and to learn from the coding style.

Every test function has a documentation  and it is helpful in differentiating the different tests for different function.

In terms of data, unlike the python package, here we have a source for the data (or at least it is easy to find). And again, Since it was sourced from a repo, is the data static or updated regularly and if yes, does your function accommodate those changes?

Again, I wish we had a little output of one or two function to directly from the README would have been great; having a section for examples to show that it works and also to allow whoever install the package to compare and see if it is the expected results.

The R package as well is great and functional ; congratulation on producing such a great package.

**Comment by robindhillon1** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples:** (that run successfully locally) for all exported functions
- [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing: 1.5

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments
- The package installed without any issues and the functions work as intended.
- 
![image](https://user-images.githubusercontent.com/86622951/216837582-97036a77-cd4a-4517-b69f-233b96c09ff2.png)

- Note, however, that there are no instructions regarding how to use the functions; in other words, there is no demonstration in the README, for example. I had to read the docstrings of the functions to use the functions. These worked as shown above. I would suggest adding a demonstration of how to use the functions. One would have to go through the roxygen documentation of the functions to check they work. This is fine, but a short demonstration is easier for everyone. See my repo [here](https://github.com/UBC-MDS/stocksignalsr) for an example.
- Similar to the python package, the hyperlink in the Dataset section in the README redirects to data that doesn't exist. Here's the link which gives me a [404 error](https://github.com/UBC-MDS/chembox/tree/main/src/chembox/data). This is an easy fix.
- There are 15 active branches in the repository. These should probably be deleted after they've been merged. 
- This package has the functions in separate scripts instead of all in one like in the python package; the same applies to the test functions. I believe this was a good decision as it increases the clarity and ease of use of these functions. Furthermore, the testing is very thorough and the code coverage is 97% which is impressive. 
- The documentation of each function is clean and easy to understand. Very pleasing to look at as well.
- The tests have appropriate comments which allow the reader to understand what exactly is being tested. This transparency is very good. 
- Similar to the python package, this is a very well-written and thorough package. To summarize, only four things I'd recommend are as follows:
    - Add short demonstration in README for new readers. 
    - Fix the 404 error
    - Delete active branches that have already been merged.
    - Add support/functionality for special isotopes that are currently not supported, such as the ones discussed in the 12-minute presentation in class.

**Comment by yukunzGIT** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
- [x] **Function Documentation:** for all exported functions
- [x] **Examples:** (that run successfully locally) for all exported functions
- [ ] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

Estimated hours spent reviewing:

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

A 97% code coverage is great. Good job, group 7.
I also have some suggestions regarding your Python package:

1. The pkgdown and vignettes links are missing from the README.
2. Maybe you could save your results data frame in a csv file for future usage.
3. For get_combustion_equation function, what if the input molecule is not combustable? Maybe try to return an error message if the user input molecule is not combustable.
4. Try to delete all unused branches after the merge to make the repo structure more clean.
5. Maybe try to add some sample usage examples in the README or at least paste the link inside RAEDME.

In summary, I really enjoy playing around with your chem package and find it could be really helpful for beginners to learn Chemistry. Great jobs, group 7. Wish you all the best!

**Comment by MNBhat** [comment]
## Package Review

Package Review
Please check off boxes as applicable, and elaborate in comments below. Your review is not limited to these topics, as described in the reviewer guide

Briefly describe any working relationship you have (had) with the package authors.
 

- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor before starting your review).

Documentation
The package includes all the following forms of documentation:
- [x] statement of need: clearly stating problems the software is designed to solve and its target audience in README
- [x] Installation instructions: for the development version of package and any non-standard dependencies in README
- [x]  Vignette(s): demonstrating major functionality that runs successfully locally
- [x] Function Documentation: for all exported functions
- [x] Examples: (that run successfully locally) for all exported functions
- [x] Community guidelines: including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with URL, BugReports and Maintainer (which may be autogenerated via Authors@R).


Functionality
- [x]  Installation: Installation succeeds as documented.
- [x]  Functionality: Any functional claims of the software been confirmed.
- [x]  Performance: Any performance claims of the software been confirmed.
 - [x] Automated tests: Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x]  Packaging guidelines: The package conforms to the rOpenSci packaging guidelines.


Estimated hours spent reviewing: 1 Hour 

 - [ ]  Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

#### Review Comments

1. The functions in this package check for the input string is correct or not. 
2.  The get_elements() function in this package does not throw an error for the different input structures which is nice to have over here. In certain scenarios like get_elements("(C2H4)"), the output has one line as 0 0 and it would be nice to remove it. 
3.  R folder has a testing_nb file, which can easily be moved to its correct place in the repo. 
4.  I would suggest removing the extra commas in the ShellConfig column in the output of the get_molec_props() function. 
5.  Some of the test functions have documentation which is great to have, it would be nice to have it in all functions. 
6. I would suggest deleting the merged branches. 

 The package is well-written and clearly states the problem and solution. Overall, Great work done. All the best !!
